### PR TITLE
Create generic tab component

### PR DESCRIPTION
**Pull request by KamWis**
**Release Type:** *Non-Breaking Feature*

Fixes https://x-team-internal.atlassian.net/browse/XP-2807

## Description

> 1. The component should hold minimal design to be effectively generic, mostly the behaviour and minimal CSS for positioning and display
> 2. Only one element can be active at a time
> 3. We should be able to inject some actions on the change for tab
> 4. Add component into Storybook and provide stories for common and edge cases

## Checklist

**Before submitting a pull request,** please make sure the following is done:

- [x] set yourself as an assignee
- [x] set appropriate labels for a PR (`In Review` or `In Progress` depending on its status)
- [x] move respective JIRA issue to the `IN REVIEW` column
- [x] make sure your code lints (`npm run lint`)
- [x] Flow typechecks passed (`npm run flow`)
- [x] Snapshots tests passed (`npm run jest`)
- [x] component's documentation (`.stories.js` file) is changed or added accordingly to reflect any new or updated use cases or variants usage
- [x] if you've fixed a bug or added code that should be tested, add unit tests
- [x] **manually tested the component** by running it in the browser and checked nothing is broken and operates as expected!

## Steps to Test or Reproduce

Go to Storybook: http://localhost:9001/?selectedKind=UI%20Components%2FTab&selectedStory=basic%20usage&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybook%2Factions%2Factions-panel&background=%23fff

## Screenshots

obraz

### CODE REVIEW

File: src/components/ui/Tab/Container.js [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+class Container extends Component<Props, State> {
+  static defaultProps = {
+    headWrapper: (props: { children: Element<*> }) => <div>{props.children}</div>,
+    contentWrapper: (props: { children: Element<*> }) => <div>{props.children}</div>,
+    onChange: () => {}
+  }
+
+  state = {
+    activeTab: this.props.defaultActiveKey || ''
```
Comment (migueliriano) [inline src/components/ui/Tab/Container.js]: If we don't have a `defaultActiveKey` we need to set the `activeTab` to the content that we are showing as default, in this case I think is the first `tab`.

the otherwise the initial active tab will work only when you set the `defaultActiveKey`

File: src/components/ui/Tab/Head.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
+// @flow
+
+import React from 'react'
+
+import type { Element } from 'react'
+
+export type Props = {
+  activeTab: string,
+  children: Element<*>,
+  onClick: Function,
```
Comment (migueliriano) [inline src/components/ui/Tab/Head.js]: Same comment here: https://github.com/x-team/auto-ui/pull/345#discussion_r256514709

File: src/components/ui/Tab/Tab.stories.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    <span
+      {...props}
+      className={`${props.className ? `${props.className}` : ''}${isActive ? ' active' : ''}`}
+    >
+      {text}
+    </span>
+  )
+}
+const HeadWrapper = (props) => <div className='header'>{props.children}</div>
+const ContentWrapper = (props) => <div className='content'>{props.children}</div>
```
[the remark this answers is not in the document]
Comment (KamWis) [reply src/components/ui/Tab/Tab.stories.js]: yep, auto-formatting happened.

File: src/components/ui/Tab/Tab.stories.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        <div>Foo bar baz 1</div>
+      </Tab.Pane>
+      <Tab.Pane tabKey='second'>
+        <div>Foo bar baz 2</div>
+      </Tab.Pane>
+      <Tab.Pane tabKey='third'>
+        <div>Foo bar baz 3</div>
+      </Tab.Pane>
+    </Tab.Container>
+  ))
```
[the remark this answers is not in the document]
Comment (KamWis) [reply src/components/ui/Tab/Tab.stories.js]: I don't see the value of making such test case - If any one implements it as a empty component - It even should and will explode. Especially in case of Head in Pane since they rely on children.

[the remark this answers is not in the document]
Comment (KamWis) [reply src/components/ui/Tab/Tab.stories.js]: Isn't that explicit that we are rendering custom component inside the tab's Head? I can add such case but I'm afraid it won't add any more value to the description. WDYT?

File: src/components/ui/Tab/Container.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+  }
+
+  handleOnChange = (activeTab: string) => {
+    this.props.onChange(activeTab)
+    this.setActiveTabOnClick(activeTab)
+  }
+
+  setActiveTabOnClick = (activeTab: string) => {
+    this.setState({ activeTab })
+  }
```
[the remark this answers is not in the document]
Comment (KamWis) [reply src/components/ui/Tab/Container.js]: yeah, good catch - It was omitted during the refactor for dirty code

File: src/components/ui/Tab/Container.js (hunk 1 above)
Comment (KamWis) [reply src/components/ui/Tab/Container.js]: ok, I will handle that

File: src/components/ui/Tab/Tab.stories.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (migueliriano) [reply src/components/ui/Tab/Tab.stories.js]: @KamWis  - I think it's ok 👍. But I just found that we don't have a case for the `onChange` event: https://github.com/x-team/auto-ui/pull/345/files#diff-7aa3fedf1591abdfd72ce51ba9daf12aR18

Can we add that story case?

[the remark this answers is not in the document]
Comment (KamWis) [reply src/components/ui/Tab/Tab.stories.js]: I thought about it and I wondered what we want to show in such case? Console.log? It's not that visible. Any suggestions? :)

[the remark this answers is not in the document]
Comment (migueliriano) [reply src/components/ui/Tab/Tab.stories.js]: Yes, for storybook we can use `action` function: https://www.npmjs.com/package/@storybook/addon-actions

This is an example where we are using it: https://github.com/x-team/auto-ui/blob/master/src/components/ui/EmailFeed.stories.js#L36

The message going to be displayed in the storybook console.

[the remark this answers is not in the document]
Comment (KamWis) [reply src/components/ui/Tab/Tab.stories.js]: cool - in this case I will add this story :) thanks!

[the remark this answers is not in the document]
Comment (bernardodiasc) [reply src/components/ui/Tab/Tab.stories.js]: > I don't see the value of making such test case - If any one implements it as an empty component - It even should and will explode.

There's actually a value on that and it should not explode. We need to make sure a wrong usage of any component doest explode the application. If corrupted data is passed on props or invalid API is used, the UI should expose that rather than make the application inoperable.

It's naive to assume that a component will be always declared correctly in the long run. This special story is an assurance agains regressions, and that we have covered the most basic declaration and that we have a visual fallback for that, even if that means doesn't render anything (as in `null` - none markup).

We do that for all UI components for these reasons, this is known as the "minimal render" test. I hope that helps clarify the value we want to achieve. :)

File: src/components/ui/Tab/Container.js [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+  onChange: (activeTab: string) => void
+}
+
+type State = {
+  activeTab: string
+}
+
+class Container extends Component<Props, State> {
+  static defaultProps = {
+    headWrapper: (props: { children: Element<*> }) => <div>{props.children}</div>,
```
Comment (bernardodiasc) [inline src/components/ui/Tab/Container.js]: Do you think the default div wrappers here are really necessary? I'm asking because:

1. I thought that the reason to create the wrapper props was to not have any predefined markup structure on this component, leaving that as a concern for the UI implementation. The behaviour without that would still work as expected, although without any visual treatment by default.

2. I had the feeling, during the call when you explained that, that you wanted to provide the most common displays for Heads and Panes, and what led you to include this was respectively the `inline` display of the `<span>` used in the showcase and the `block` display you expected by default. But I guess it was for the demo to not look strange.

I played a bit with this component without the default `<div>` wrappers using instead:
```
    headWrapper: (props: { children: Element<*> }) => <Fragment>{props.children}</Fragment>,
    contentWrapper: (props: { children: Element<*> }) => <Fragment>{props.children}</Fragment>,
```

And instead using a `<span>` I pick a `<Button>`. And the Pane contents was using a `<div>`. And included some colours and spacing for custom wrappers using `<div>`. I tried a bunch of alternatives and all seemed to go pretty fine without default wrapper markup.

```
    <Tab.Container headWrapper={HeadWrapper} contentWrapper={ContentWrapper}>
      <Tab.Head tabKey='first'>
        <ExampleButton text='Tab 1' className='foo' />
      </Tab.Head>
      <Tab.Head tabKey='second'>
        <ExampleButton text='Tab 2' className='foo' />
      </Tab.Head>
      <Tab.Head tabKey='third'>
        <ExampleButton text='Tab 3' className='bar' />
      </Tab.Head>
      <Tab.Pane tabKey='first'>
        <div>Foo bar baz 1</div>
      </Tab.Pane>
      <Tab.Pane tabKey='second'>
        <div>Foo bar baz 2</div>
      </Tab.Pane>
      <Tab.Pane tabKey='third'>
        <div>Foo bar baz 3</div>
      </Tab.Pane>
    </Tab.Container>
```

Using custom `headWrapper` and `contentWrapper`:


Using only custom `headWrapper`:


Using only custom `contentWrapper`:


Using **none** wrappers (all tab heads and pane here are sibling elements):



---

Bottom line, we can safely simplify the markup output that way. Even on those examples I could use a styled component instead `<div>Foo bar baz 1</div>` and that would make the `contentWrapper` really optional for the demos above.

Comment (bernardodiasc) [reply src/components/ui/Tab/Container.js]: Also, the current showcases are good ones. After a discussion with Paul we realised that sometimes the showcases demonstrates the functionalities but have low value for designers or even him. It's the case of this component because it's not delivering any visual by default.

For that reason I'd suggest to prepare a story with a fancy custom UI, to expose the possibilities of UI customisation. That can be a lot useful also to make sure it's flexible enough for many different use cases.

But of course, if you think this is out of the scope, we address that on a separated cleanup task when we'll be replacing this new Tab in the other places in the app. That will be good opportunity to put to the test the flexibility of a generic Tab. :)

Comment (KamWis) [reply src/components/ui/Tab/Container.js]: thanks Bernardo :) So in the end I will replace those `div`s with the `Fragment` goods idea!

Comment (KamWis) [reply src/components/ui/Tab/Container.js]: Regarding fancy stories - I will create a ticket for it - I suggest that it shouldn't be a clean-up ticket but a user story - the reason for it is that it's touching our user and provides tool for the designer, so it's not only impacting us. WDYT?

File: src/components/ui/Tab/Tab.stories.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (KamWis) [reply src/components/ui/Tab/Tab.stories.js]: fair enough - will sanitize case for lack of children. Tanks m8 :)

**Comment by KamWis** [comment]
@bernardodiasc and @migueliriano  - all your comments has been addressed.
Latest update: scenario for missing props for Container, Head and Pane has been added and default wrapper has been change to use `Fragment` instead of `div`

File: src/components/ui/Tab/Container.js (hunk 6 above)
Comment (bernardodiasc) [reply src/components/ui/Tab/Container.js]: Sounds good to me 👍
